base/inference.jl

```diff
@@ -2607,21 +2607,12 @@ function tmerge(@nospecialize(typea), @nospecialize(typeb))
     if !(isa(typea,Type) || isa(typea,TypeVar)) || !(isa(typeb,Type) || isa(typeb,TypeVar))
         return Any
     end
-    if (typea <: Tuple) && (typeb <: Tuple)
-        if isa(typea, DataType) && isa(typeb, DataType) && length(typea.parameters) == length(typeb.parameters) && !isvatuple(typea) && !isvatuple(typeb)
-            return typejoin(typea, typeb)
-        end
-        if isa(typea, Union) || isa(typeb, Union) || (isa(typea,DataType) && length(typea.parameters)>3) ||
-            (isa(typeb,DataType) && length(typeb.parameters)>3)
-            # widen tuples faster (see #6704), but not too much, to make sure we can infer
-            # e.g. (t::Union{Tuple{Bool},Tuple{Bool,Int}})[1]
-            return Tuple
-        end
-    end
     u = Union{typea, typeb}
-    if unionlen(u) > MAX_TYPEUNION_LEN || type_too_complex(u, MAX_TYPE_DEPTH)
-        # don't let type unions get too big
-        # TODO: something smarter, like a common supertype
+    if unionlen(u) > MAX_TYPEUNION_LEN
+        u = typejoin(typea, typeb)
```

Comment:
Not sure I completely understand the problem... if `typea <: typeb` or vice verse, we don't get here. Now if say `typea` is a `Union`, surely neither `type.a <: typea.b` nor vice verse, so in the [recursive call](https://github.com/JuliaLang/julia/blob/fb76529050cdf142ce6614b28d690fff560e8d3a/base/promotion.jl#L22-L23) of `typejoin`, the subtype test should always be false. Likewise for `typeb`, of course. But admittedly, if say `typea` is a `TypeVar` and `typeb` is a `Union`, things get a little trickier to reason about. So no matter whether it's actually necessary, your proposed version is much easier to reason about :)